File: bindings/java/src/test/java/org/wickra/backtest/StreamingBacktestTests.java
package org.wickra.backtest;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import org.json.JSONArray;
import org.json.JSONObject;
import org.junit.jupiter.api.Test;

/**
 * The streaming class must be the same engine as {@link Backtester#run}, one bar
 * at a time -- that equivalence is the claim, so it is what these tests pin.
 */
class StreamingBacktestTests {

    private static final String PRICE_SPEC =
            "{\"symbol\":\"x\",\"timeframe\":\"1h\",\"indicators\":{},"
            + "\"entry\":{\"gt\":[{\"price\":\"close\"},100]},"
            + "\"exit\":{\"lt\":[{\"price\":\"close\"},100]},"
            + "\"sizing\":{\"type\":\"fixed_qty\",\"qty\":1}}";

    private static final double[] OPEN = {100.0, 102.0, 104.0, 98.0};
    private static final double[] HIGH = {101.0, 103.0, 104.0, 98.0};
    private static final double[] LOW = {100.0, 102.0, 99.0, 97.0};
    private static final double[] CLOSE = {101.0, 103.0, 99.0, 97.0};

    private static String batchReport() {
        return Backtester.run(OPEN, HIGH, LOW, CLOSE, PRICE_SPEC, 1000.0);
    }

    @Test
    void streamingReproducesTheBatchReport() {
        try (StreamingBacktest bt = new StreamingBacktest(PRICE_SPEC, 1000.0)) {
            for (int i = 0; i < OPEN.length; i++) {
                bt.step(OPEN[i], HIGH[i], LOW[i], CLOSE[i]);
            }
            assertEquals(batchReport(), bt.finishJson());
        }
    }

    @Test
    void stepJsonMatchesTheScalarStep() {
        try (StreamingBacktest bt = new StreamingBacktest(PRICE_SPEC, 1000.0)) {
            for (int i = 0; i < OPEN.length; i++) {
                bt.stepJson("{\"candle\":{\"time\":" + i
                        + ",\"open\":" + OPEN[i] + ",\"high\":" + HIGH[i]
                        + ",\"low\":" + LOW[i] + ",\"close\":" + CLOSE[i]
                        + ",\"volume\":0}}");
            }
            assertEquals(batchReport(), bt.finishJson());
        }
    }

    @Test
    void accessorsTrackTheRun() {
        try (StreamingBacktest bt = new StreamingBacktest(PRICE_SPEC, 1000.0)) {
            assertEquals("null", bt.latestEquityJson());
            assertEquals(0, new JSONArray(bt.equityJson()).length());
            assertEquals(0L, bt.numTrades());
            assertFalse(bt.isFinished());

            for (int i = 0; i < 3; i++) {
                bt.step(OPEN[i], HIGH[i], LOW[i], CLOSE[i]);
            }

            JSONArray curve = new JSONArray(bt.equityJson());
            assertEquals(3, curve.length());
            // Bar 2 closed below 100, which is the exit *signal*; the fill lands
            // on the next bar's open, so nothing has closed yet.
            assertEquals(0L, bt.numTrades());

            bt.step(OPEN[3], HIGH[3], LOW[3], CLOSE[3]);
            assertEquals(1L, bt.numTrades());
        }
    }

    @Test
    void stepDefaultsTheTimestampToTheBarIndex() {
        try (StreamingBacktest bt = new StreamingBacktest(PRICE_SPEC, 1000.0)) {
            for (int i = 0; i < OPEN.length; i++) {
                bt.step(OPEN[i], HIGH[i], LOW[i], CLOSE[i]);
            }
            JSONArray curve = new JSONArray(bt.equityJson());
            assertEquals(OPEN.length, curve.length());
            for (int i = 0; i < curve.length(); i++) {
                assertEquals(i, curve.getJSONObject(i).getLong("time"));
            }
        }
    }

    @Test
    void aFinishedRunRefusesFurtherUse() {
        StreamingBacktest bt = new StreamingBacktest(PRICE_SPEC, 1000.0);
        bt.step(OPEN[0], HIGH[0], LOW[0], CLOSE[0]);
        bt.finishJson();
        assertTrue(bt.isFinished());

        assertThrows(IllegalStateException.class, () -> bt.step(1.0, 1.0, 1.0, 1.0));
        assertThrows(IllegalStateException.class, () -> bt.stepJson("{}"));
        assertThrows(IllegalStateException.class, bt::equityJson);
        assertThrows(IllegalStateException.class, bt::latestEquityJson);
        assertThrows(IllegalStateException.class, bt::numTrades);
        assertThrows(IllegalStateException.class, bt::finishJson);
    }

    @Test
    void closeIsIdempotentAndComposesWithFinish() {
        StreamingBacktest bt = new StreamingBacktest(PRICE_SPEC, 1000.0);
        bt.step(OPEN[0], HIGH[0], LOW[0], CLOSE[0]);
        bt.close();
        bt.close();
        assertTrue(bt.isFinished());

        // finishJson already released the handle, so the try-with-resources
        // close that follows it must be a no-op rather than a double free.
        try (StreamingBacktest other = new StreamingBacktest(PRICE_SPEC, 1000.0)) {
            other.step(OPEN[0], HIGH[0], LOW[0], CLOSE[0]);
            other.finishJson();
        }
    }

    @Test
    void anInvalidSpecThrows() {
        assertThrows(IllegalStateException.class,
                () -> new StreamingBacktest("{\"bad\":true}", 1000.0));
    }

    /**
     * A pairwise indicator is undefined without its reference series, so a spec
     * that reads one proves the per-bar feed actually arrives -- and it must
     * agree with the batch path fed the same reference.
     */
    @Test
    void perBarFeedsReachAReferenceReadingStrategy() {
        // A sine path, not a geometric one: constant growth means constant log
        // returns, which drives the correlation's variance to zero.
        int n = 24;
        double[] closes = new double[n];
        for (int i = 0; i < n; i++) {
            closes[i] = 100.0 + 10.0 * Math.sin(i * 0.5);
        }
        String spec = "{\"symbol\":\"x\",\"timeframe\":\"1h\","
                + "\"indicators\":{\"corr\":{\"type\":\"PearsonCorrelation\",\"params\":[5]}},"
                + "\"entry\":{\"gt\":[\"corr\",0.5]},\"exit\":{\"lt\":[\"corr\",-0.5]},"
                + "\"sizing\":{\"type\":\"fixed_qty\",\"qty\":1}}";

        StringBuilder candles = new StringBuilder();
        StringBuilder reference = new StringBuilder();
        String streamed;
        try (StreamingBacktest bt = new StreamingBacktest(spec, 1000.0)) {
            for (int i = 0; i < n; i++) {
                double c = closes[i];
                double ref = 2.0 * c;
                String candle = "{\"time\":" + i + ",\"open\":" + c + ",\"high\":" + (c + 1.0)
                        + ",\"low\":" + (c - 1.0) + ",\"close\":" + c + ",\"volume\":0}";
                if (i > 0) {
                    candles.append(',');
                    reference.append(',');
                }
                candles.append(candle);
                reference.append("{\"time\":").append(i).append(",\"open\":").append(ref)
                        .append(",\"high\":").append(ref).append(",\"low\":").append(ref)
                        .append(",\"close\":").append(ref).append(",\"volume\":0}");
                bt.stepJson("{\"candle\":" + candle + ",\"feeds\":{\"reference\":" + ref + "}}");
            }
            streamed = bt.finishJson();
        }

        String batch = Backtester.runJson("{\"spec\":" + spec + ",\"capital\":1000,\"candles\":["
                + candles + "],\"reference\":[" + reference + "]}");
        assertEquals(batch, streamed);
        assertEquals(1, trades(streamed));

        // The feed is load-bearing: without it the correlation never resolves.
        String blind;
        try (StreamingBacktest bt = new StreamingBacktest(spec, 1000.0)) {
            for (int i = 0; i < n; i++) {
                double c = closes[i];
                bt.step(c, c + 1.0, c - 1.0, c);
            }
            blind = bt.finishJson();
        }
        assertEquals(0, trades(blind));
        assertNotEquals(streamed, blind);
    }

    private static int trades(String reportJson) {
        return new JSONObject(reportJson).getJSONObject("metrics").getInt("num_trades");
    }
}

File: bindings/csharp/Wickra.Backtest.Tests/StreamingBacktestTests.cs
using System;
using System.Text.Json;
using Wickra.Backtest;
using Xunit;

// The streaming class must be the same engine as Backtester.Run, one bar at a
// time -- that equivalence is the claim, so it is what these tests pin.
public class StreamingBacktestTests
{
    private const string PriceSpec =
        "{\"symbol\":\"x\",\"timeframe\":\"1h\",\"indicators\":{}," +
        "\"entry\":{\"gt\":[{\"price\":\"close\"},100]}," +
        "\"exit\":{\"lt\":[{\"price\":\"close\"},100]}," +
        "\"sizing\":{\"type\":\"fixed_qty\",\"qty\":1}}";

    private static readonly double[] Open = { 100.0, 102.0, 104.0, 98.0 };
    private static readonly double[] High = { 101.0, 103.0, 104.0, 98.0 };
    private static readonly double[] Low = { 100.0, 102.0, 99.0, 97.0 };
    private static readonly double[] Close = { 101.0, 103.0, 99.0, 97.0 };

    private static string BatchReport() =>
        Backtester.Run(Open, High, Low, Close, spec: PriceSpec, capital: 1000.0);

    [Fact]
    public void StreamingReproducesTheBatchReport()
    {
        using var bt = new StreamingBacktest(PriceSpec, 1000.0);
        for (int i = 0; i < Open.Length; i++)
        {
            bt.Step(Open[i], High[i], Low[i], Close[i]);
        }
        Assert.Equal(BatchReport(), bt.FinishJson());
    }

    [Fact]
    public void StepJsonMatchesTheScalarStep()
    {
        using var bt = new StreamingBacktest(PriceSpec, 1000.0);
        for (int i = 0; i < Open.Length; i++)
        {
            string doc =
                "{\"candle\":{\"time\":" + i +
                ",\"open\":" + Open[i] + ",\"high\":" + High[i] +
                ",\"low\":" + Low[i] + ",\"close\":" + Close[i] + ",\"volume\":0}}";
            bt.StepJson(doc);
        }
        Assert.Equal(BatchReport(), bt.FinishJson());
    }

    [Fact]
    public void AccessorsTrackTheRun()
    {
        using var bt = new StreamingBacktest(PriceSpec, 1000.0);
        Assert.Equal("null", bt.LatestEquityJson());
        Assert.Equal("[]", bt.EquityJson());
        Assert.Equal(0, bt.NumTrades);
        Assert.False(bt.IsFinished);

        for (int i = 0; i < 3; i++)
        {
            bt.Step(Open[i], High[i], Low[i], Close[i]);
        }

        using var curve = JsonDocument.Parse(bt.EquityJson());
        Assert.Equal(3, curve.RootElement.GetArrayLength());
        // Bar 2 closed below 100, which is the exit *signal*; the fill lands on
        // the next bar's open, so nothing has closed yet.
        Assert.Equal(0, bt.NumTrades);

        bt.Step(Open[3], High[3], Low[3], Close[3]);
        Assert.Equal(1, bt.NumTrades);
    }

    [Fact]
    public void TimeDefaultsToTheBarIndex()
    {
        using var bt = new StreamingBacktest(PriceSpec, 1000.0);
        for (int i = 0; i < Open.Length; i++)
        {
            bt.Step(Open[i], High[i], Low[i], Close[i]);
        }
        using var curve = JsonDocument.Parse(bt.EquityJson());
        int index = 0;
        foreach (var point in curve.RootElement.EnumerateArray())
        {
            Assert.Equal(index, point.GetProperty("time").GetInt64());
            index++;
        }
        Assert.Equal(Open.Length, index);
    }

    [Fact]
    public void AFinishedRunRefusesFurtherUse()
    {
        var bt = new StreamingBacktest(PriceSpec, 1000.0);
        bt.Step(Open[0], High[0], Low[0], Close[0]);
        bt.FinishJson();
        Assert.True(bt.IsFinished);

        Assert.Throws<ObjectDisposedException>(() => bt.Step(Open[1], High[1], Low[1], Close[1]));
        Assert.Throws<ObjectDisposedException>(() => bt.EquityJson());
        Assert.Throws<ObjectDisposedException>(() => bt.LatestEquityJson());
        Assert.Throws<ObjectDisposedException>(() => bt.FinishJson());
        Assert.Throws<ObjectDisposedException>(() => bt.NumTrades);
    }

    [Fact]
    public void DisposeIsIdempotentAndEndsTheRun()
    {
        var bt = new StreamingBacktest(PriceSpec, 1000.0);
        bt.Step(Open[0], High[0], Low[0], Close[0]);
        bt.Dispose();
        bt.Dispose();
        Assert.True(bt.IsFinished);
    }

    [Fact]
    public void AnInvalidSpecThrows()
    {
        Assert.Throws<InvalidOperationException>(() => new StreamingBacktest("{\"bad\":true}"));
    }

    [Fact]
    public void PerBarFeedsReachAReferenceReadingStrategy()
    {
        // A sine path, not a geometric one: constant growth means constant log
        // returns, which drives the correlation's variance to zero.
        const int N = 24;
        var closes = new double[N];
        for (int i = 0; i < N; i++)
        {
            closes[i] = 100.0 + 10.0 * Math.Sin(i * 0.5);
        }

        const string Spec =
            "{\"symbol\":\"x\",\"timeframe\":\"1h\"," +
            "\"indicators\":{\"corr\":{\"type\":\"PearsonCorrelation\",\"params\":[5]}}," +
            "\"entry\":{\"gt\":[\"corr\",0.5]},\"exit\":{\"lt\":[\"corr\",-0.5]}," +
            "\"sizing\":{\"type\":\"fixed_qty\",\"qty\":1}}";

        string streamed;
        using (var bt = new StreamingBacktest(Spec, 1000.0))
        {
            for (int i = 0; i < N; i++)
            {
                double c = closes[i];
                bt.StepJson(
                    "{\"candle\":{\"time\":" + i + ",\"open\":" + c +
                    ",\"high\":" + (c + 1.0) + ",\"low\":" + (c - 1.0) +
                    ",\"close\":" + c + ",\"volume\":0}," +
                    "\"feeds\":{\"reference\":" + (2.0 * c) + "}}");
            }
            streamed = bt.FinishJson();
        }

        using var streamedDoc = JsonDocument.Parse(streamed);
        Assert.Equal(
            1,
            streamedDoc.RootElement.GetProperty("metrics").GetProperty("num_trades").GetInt32());

        // The feed is load-bearing: without it the correlation never resolves,
        // so the strategy never fires and the two runs cannot agree.
        string blind;
        using (var bt = new StreamingBacktest(Spec, 1000.0))
        {
            for (int i = 0; i < N; i++)
            {
                double c = closes[i];
                bt.Step(c, c + 1.0, c - 1.0, c);
            }
            blind = bt.FinishJson();
        }

        using var blindDoc = JsonDocument.Parse(blind);
        Assert.Equal(
            0,
            blindDoc.RootElement.GetProperty("metrics").GetProperty("num_trades").GetInt32());
        Assert.NotEqual(blind, streamed);
    }
}
